### DISCUSSION TITLE
Sorting dataset based on the nodes' page rank values

### DISCUSSION THREAD

**Discussion by OliverSummers** [pyg-team/pytorch_geometric#3004] [Q&A]
I asked a similar question in another post, and according to the provided answer to sort the Planetoid dataset based on nodes' `PageRank`, I coded as follows:

 
     import networkx as nx
     from torch_geometric.utils.convert import to_networkx
     from torch_geometric.datasets import Planetoid
     
     dataset = Planetoid('./data', 'PubMed')
     data = dataset[0]

     nx_data = to_networkx(data)
     ranks = nx.pagerank(nx_data, alpha=0.9, tol=1.e-08)
     pagerank = list(ranks.values())
     pagerank_tensor = torch.tensor(pagerank)
     perm = pagerank_tensor.argsort()

     inv_perm = torch.empty_like(perm)
     inv_perm[perm] = torch.arange(perm.size(0))

     new_x = data.x[perm]
     new_y = data.y[perm]
     new_edge_index = inv_perm[data.edge_index]

     data.x = new_x
     data.y = new_y
     data.edge_index = new_edge_index

     dataset.transform = data

Then I check the transformed dataset to see if it is sorted based on the nodes' PageRank. The following snippet prints out the indices of the consecutive nodes with the same `PageRank` values:

     trans_data = dataset[0]
     nx_graph = to_networkx(trans_data)
     ranks = nx.pagerank(nx_graph, alpha=0.9, tol=1.e-08)
     rank_list = list(ranks.values())

     s = set(rank_list)
     s_list = list(s)
     dict_inx = {}

     for i in s_list:
        first_index = ran_list.index(i)

        reversed_list = ran_list[::-1]

        first_index_in_reversed = reversed_list.index(i)

        last_index = len(ran_list) - 1 - first_index_in_reversed

        dict_inx[i] = (first_index, last_index)
     
     for k, v in dict_inx.items():
       print(k, v)

The output of the above snippet is:

      1.5705199615468062e-05 : (3339, 3340)

      5.347380017895053e-05 : (15102, 15102)

      2.597480481370294e-05 : (10631, 10634)

      .

      .

      .

      2.672946705217494e-05 : (10935, 10935)

      2.771347590169145e-05 : (11333, 11333)

where the number before `:` is `PageRank` and the tuple after `:` denotes indices of all nodes with the same `PageRank`. 
As we can see, the page ranks are sorted in ascending order but their corresponding indices are not sorted. If all of the above code is written based on the nodes' `degree` values then the final output will be:

     1 : (0, 9093)

     2 : (9094, 12450)

     3 : (12451, 14034)

     .
   
     .
  
     .

     154 : (19715, 19715)

     171 : (19716, 19716)`

where the number before `:` is `degree` and the tuple after `:` refers to the nodes' indices with the same `degree` values. 
So as we can see, the results with the sorted dataset based on `degree` values are perfect. Both degrees and indices are sorted ascendingly but as I explained above, it is not the case with `PageRank`.
My question is how to fix the problem of the `PageRank` code?

**Comment by rusty1s** [comment]
Sorting of the inner lists is not really necessary, but if you want to do that I suggest iterating over each page rank value and sorting the list of node indices manually, e.g.,:
```python
for k, v in dict_inx.items():
   dict_inx[k] = sorted(v)
```

**Reply by OliverSummers** [reply]
`v` tuple is already sorted but my point was how to sort the tuples one after the other. For example in sorting based on degree, the nodes with degree 1 correspond to index 0 to **9093**, and then with degree 2 it starts at **9094** to 12450 however in PageRank case, in the transformed dataset, the smallest page rank does not start at index 0 but at index 3339 and then the next page rank starts at index 15102. I hope I could clarify my question better.

**Reply by rusty1s** [reply]
I don't see a problem here. Simply converting/flatten your dictionary into a list sorted based on PageRank value
will give you a similar vector as what you would have received after calling `degree.argmax()`, i.e., the permutation of the nodes so that they are sorted by PageRank value.
